src/Controller/SiteInfoController.php

```diff
@@ -0,0 +1,203 @@
+   *   The entity field manager service.
+   */
+  public function __construct(SamplerPluginManager $sampler_plugin_manager, Mapping $sampler_mapping, EntityFieldManagerInterface $entity_field_manager) {
+    $this->samplerPluginManager = $sampler_plugin_manager;
+    $this->samplerMapping = $sampler_mapping;
+    $this->entityFieldManager = $entity_field_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('plugin.manager.sampler'),
+      $container->get('sampler.mapping'),
+      $container->get('entity_field.manager')
+    );
+  }
+
+  /**
+   * Get required fields for bundle.
+   *
+   * @param string $entity_type
+   *   The entity type.
+   * @param string $bundle
+   *   The bundle name.
+   *
+   * @return mixed
+   *   Returns all required fields for bundle.
+   */
+  protected function getRequiredFieldWidgets($entity_type, $bundle) {
+    $entity_form_display = EntityFormDisplay::load("{$entity_type}.{$bundle}.default");
+    $form_display_widgets = $entity_form_display->getComponents();
+
+    $field_definitions = $this->entityFieldManager->getFieldDefinitions($entity_type, $bundle);
+
+    $required_fields = array_reduce(
+      array_keys($form_display_widgets),
+      function ($collection, $field_name) use ($form_display_widgets, $field_definitions) {
+        $field_info = $form_display_widgets[$field_name];
+        if (!isset($field_info['region']) || $field_info['region'] !== 'content') {
+          return $collection;
+        }
+
+        if (isset($field_definitions[$field_name]) && $field_definitions[$field_name]->isRequired()) {
+          $collection[$field_name] = $field_info;
+        }
+
+        return $collection;
+      },
+      []
+    );
+
+    return $required_fields;
+  }
+
+  /**
+   * Get bundles ordered by number of instances.
+   *
+   * @param array $data
+   *   The collected data from sampler module for entity type.
+   *
+   * @return array
+   *   Returns bundle names ordered by count.
+   */
+  protected function getBundlesByCount(array $data) {
+    $bundles_by_count = [];
+
+    foreach ($data as $bundle_name => $bundle_info) {
+      if (!isset($bundle_info['instances'])) {
+        continue;
+      }
+
+      $bundles_by_count[$bundle_name] = $bundle_info['instances'];
+    }
+
+    arsort($bundles_by_count);
+
+    return $bundles_by_count;
+  }
+
+  /**
+   * Get bundles ordered by number of instances.
+   *
+   * @param array $data
+   *   The collected data from sampler module for entity type.
+   *
+   * @return array
+   *   Returns bundle names ordered by count.
+   */
+  protected function getBundlesByNumberOfFields(array $data) {
+    $bundles_by_number_of_fields = [];
+
+    foreach ($data as $bundle_name => $bundle_info) {
+      $bundles_by_number_of_fields[$bundle_name] = count(array_keys($bundle_info['fields']));
```

Comment:
The $data array contains the 'bundles_by_count' key at this point, which does not have the 'fields' values.
This throws some notices. We either test, if the 'fields' key actually exists, or do not store the result of the getBundlesBy*() results in the data array.